app/assets/javascripts/angular/controllers/graph_ctrl.js

```diff
@@ -123,10 +126,15 @@ angular.module("Prometheus.controllers").controller('GraphCtrl',
       var rangeSeconds = Prometheus.Graph.parseDuration(scope.graph.range);
       // bigger denominator == smaller step == more data
       var step = Math.floor(rangeSeconds / scalingFactor)
-      refreshFn(rangeSeconds, step).then(function(data) {
+
+      // Cancels the reload request if it exists.
+      $timeout.cancel(debounce);
+      debounce = $timeout(function() {
+        refreshFn(rangeSeconds, step).then(function(data) {
         scope.$broadcast('redrawGraphs', data);
         AnnotationRefresher(scope.graph, scope);
-      });
+        });
+      }, 500);
```

Comment:
What concerns me is that this 500ms debounce is now added to any and all graph loads, not just the ones that are triggered by time adjustments. This includes initial page load, as well as any other manual changes to graphs which don't profit from being debounced. One solution I can think of is to pass a boolean `debounce` argument to `refreshGraph()`, which would only be set to `true` when the refresh originates from a time settings change, namely in these lines:

https://github.com/prometheus/promdash/blob/debounce/app/assets/javascripts/angular/controllers/graph_ctrl.js#L94-L102